File: in_clojure/day_02_part_1.clj
(require '[clojure.string :as str])

(def input
  (-> "../inputs/day02.txt"
      (slurp)
      (str/trim)
      (str/split #",")
      (->> (mapv read-string))))

(def ops {1 +, 2 *, 99 :return})

(defn execute [program-init]
  (loop [pos 0 program program-init]
    (let [op-id (get program pos)
          op (get ops op-id)]
      (if (= op :return)
        program
        (let [a*   (get program (+ 1 pos))
              b*   (get program (+ 2 pos))
              res* (get program (+ 3 pos))
              res  (op (get program a*)
                       (get program b*))]
          (recur (+ pos 4) (assoc program res* res)))))))

(def result
  (first
   (execute (assoc input
                   1 12
                   2 2))))

(println "Day 02 - Part 1:" result)

File: in_ruby/day_02_part_1.rb
input = File.read("../inputs/day02.txt").split(",").map(&:to_i)

ADD=1
MULT=2
RETURN=99

def execute!(prog)
  prog = prog.clone
  pos = 0
  loop do
    op_id,a_star,b_star,res_star = prog[pos .. pos+3]
    break if op_id == RETURN
    op = case op_id
         when ADD  then :+.to_proc
         when MULT then :*.to_proc
         end
    prog[res_star] = op[prog[a_star], prog[b_star]]
    pos+=4
  end
  prog
end

input[1]=12
input[2]=2
result = execute!(input)

puts "Day 02 - Part 1: #{result[0]}"
